### DISCUSSION TITLE
BG: first USECODE version deletes inn keys when sleeping

### DISCUSSION THREAD

**Discussion by DominusExult** [exult/exult#380] [General]
The first release of BG deletes the inn keys when you sleep. This was later *fixed* by BG version 3.4. That patch does that by replacing the bytes `0x0D21` with `0x001F` at offset `0x1117CF` of the USECODE file.

@marzojr  investigated and this hacks the usecode by turning
`if (!var000D) goto labelFunc0622_014D;`
into
`if (!0x0000) goto labelFunc0622_014D;`
This neutralizes a loop that was meant to remove keys of any frame with quality 255.

Exult follows what the usecode does so with that first USECODE version, we delete the keys as well.

The original idea was good but not completely thought through as this imprisons the party in the inn bedroom when they locked the door before going to sleep.
IMO a way around that would have been to remove the keys but also unlock all doors with the quality 255 in a small radius. And then maybe automatically lock them again on cache out/in. Way more elaborate and the danger of locking doors that aciddentally have that quality.

No idea if we should circumvent that for the unlikely case that someone uses the original version...

We found that by looking at the 3.4 patch (https://archive.org/details/U7UP34_ZIP).
It consists of 4 files:
UPDATEU7.BAT (which starts patching, usage: UPDATEU7.BAT PATH_TO_U7)
KZ.EXE
U7.KZ
XXX.EXE

KZ.EXE unpacks U7.KZ to U7.EXE (version 3.4)
XXX.EXE patches the USECODE file as detailed above. Curiosly, if you use XXX.EXE directly (XXX.EXE PATH_TO_U7) with an unpatched 3.0 version of BG it will also patch U7.exe by just changing its version number to 3.1.

**Comment by marzojr** [comment]
It is worth noting that the FoV usecode fixes this in a different way: it removes the loop entirely.
